rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,226 @@
+layout. To switch to federated storage, the table will need to be re-bootstrapped with the new layout.
+
+3. The Instant metadata (`HoodieCommitMetadata`,`HoodieCleanMetadata` etc.) will always act as the source of file listing
+for metadata table to be populated.
+
+4. `HoodieCommitMetadata` currently stores `file name` instead of complete `file path`. We will have to modify commit
+metadata to store the complete file path instead of just file name, as the files are now distributed across several random
+prefix paths instead of a derivable table/partition path.
+
+5. If there is an error reading from Metadata table, we will not fall back listing from file system.
+
+6. In case of metadata table getting corrupted or lost, we need to have a solution here to reconstruct metadata table
+from the files which distributed using federated storage. We will likely have to implement a file system listing
+logic, that can get all the partition to files mapping by listing all the prefixes under the `Table Storage Path`.
+Following the folder structure of adding table name/partitions under the prefix will help in getting the listing and
+identifying the table/partition they belong to.
+
+### Query Side Integration
+
+Spark, Hive and [Presto](https://github.com/prestodb/presto/commit/ef1fd25c582631513ccdd097e0a654cda44ec3dc) are
+already integrated to use metadata based listing. Hudi connector for [Trino](https://github.com/prestodb/presto/commit/ef1fd25c582631513ccdd097e0a654cda44ec3dc)
+is currently in process of being contributed, which also integrates with metadata based listing. In general, since these
+query engines are able to use Hudi's metadata table there should ideally be no changes required in
+terms of making them work with federated storage. Here are some considerations:
+
+- Spark DataSource and Spark SQL queries have been integrated with metadata based listing via the Hudi's custom implementation
+of Spark's FileIndex interface. However, if Spark DataSource queries are used with globbed paths then the FileIndex path
+does not kick in, and it would rely on Spark's `InMemoryFileIndex` to do the file listing with Hudi's path filter applied. Thus,
+these Spark DataSource queries with globbed paths would not work with federated storage.
+
+- Query engines should be able to determine that federated storage is configured, and rely on metadata to list files. It
+should not be user's responsibility to enable metadata listing from query engines side.
+
+- We need to ensure that partition pruning continues to work for the query engines.
+
+### Future Work
+
+- Currently, we are only targeting Hudi `Data files` to be stored in federated storage format. However, Hudi also
+generates `Marker files` corresponding to each data file which sits under the Hudi metadata folder. It can still be a
+bottleneck as all the marker files will be stored under the table path prefix.
```

Comment:
The [timeline-server-based markers](https://hudi.apache.org/blog/2021/08/18/improving-marker-mechanism/) has already solved the problem of storing each individual marker as a single file under `.temp/`.

Reply:
Good point, I will take a look